documentation.md

```diff
@@ -0,0 +1,669 @@
+The oldest candidate within this cost category
+
+Additionally, the cost and row buffer identifier corresponding to the optimal entry are determined (resp. on the signals _opti_cost_cat_ and _opti_rbuf_).
+This row buffer identifier helps updating the simulated rank state.
+
+### Detailed slot operation
+
+#### Request acceptance
+
+##### Write requests
+
+When a write slot is free (_~v_), a new write address request can be accepted.
+The allocated free slot fields takes the following values:
+
+- _v_ is set.
+- _iid_ takes the value held by the _waddr_iid_i_ signal, which is the write reservation output signal from the response banks.
+- _addr_ takes the value held by the address request.
+- _burst_size_ takes the value held by the address request.
+- _burst_fixed_ takes the value 1 iff the burst type is fixed.
+- _data_v_ takes the value, for each bit:
+  - 1 for the first _wdata_immediate_cnt_i_ bits, which correspond to the number of write data requests that had arrived not later than the write address request.
+  - 0 for the next bits corresponding to entries in the burst.
+  - 1 for the last entries, which correspond to entries beyond the burst length.
+- _mem_pending_ takes the value '0, as no simulated memory operation has started for any of these new entries.
+- _mem_done_ is set to zero, except for the last bits, which correspond to entries beyond the burst length.
+  The latter bits are set, because an unset bit in the _mem_done_ array represents an entry which must be eventually completed.
+
+<figure class="image">
+  <img src="https://i.imgur.com/eUkYbI9.png" alt="Write slot state">
+  <figcaption>Fig: Write slot state just after acceptance of a new write address of burst length 2 and 1 immediate wdata (write data arrived not after the write address request)</figcaption>
+</figure>
+
+##### Read requests
+
+Read request acceptance is identical to write address acceptance, except that:
+
+- Read and write slots are disjoint.
+- The _data_v_ field is absent from read slots.
+  Virtually, all the bits would be set to 1 when a read slot is allocated.
+
+#### Entry and slot liberation
+
+For each write or read slot, for each entry, if the corresponding _mem_pending_ bit is set, then it is swapped with the _mem_done_ bit when the corresponding rank decrementing counter reaches 3.
+This accommodates the propagation delay until the requester, assuming the latter is ready.
+This is referred as _three-cycles-early_ _mem_done_ setting.
+
+It forces, in particular, all the simulated memory delays to be at least 3 cycles.
+This lower bound is much lower than typical main memory access delays.
+
+##### Write requests
+
+As there is a single write response per write address request, the release of the write response corresponding to write slot is enabled when all the entries (_i.e._, all the write data requests) of the slot are completed: when the _mem_done_ array is full of ones.
```

Comment:
"...are completed. This is the case if _mem_done_ is full of ones."

Reply:
Done!